lib/datadog/core/crashtracking/agent_base_url.rb

```diff
@@ -7,10 +7,17 @@ module Core
     module Crashtracking
       # This module provides a method to resolve the base URL of the agent
       module AgentBaseUrl
+        # IPv6 regular expression from
+        # https://stackoverflow.com/questions/53497/regular-expression-that-matches-valid-ipv6-addresses
+        # Does not match IPv4 addresses.
+        IPV6_REGEXP = /\A(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\z)/.freeze # rubocop:disable Layout/LineLength
+
         def self.resolve(agent_settings)
           case agent_settings.adapter
           when Datadog::Core::Configuration::Ext::Agent::HTTP::ADAPTER
-            "#{agent_settings.ssl ? 'https' : 'http'}://#{agent_settings.hostname}:#{agent_settings.port}/"
+            hostname = agent_settings.hostname
+            hostname = "[#{hostname}]" if hostname =~ IPV6_REGEXP
+            "#{agent_settings.ssl ? 'https' : 'http'}://#{hostname}:#{agent_settings.port}/"
           when Datadog::Core::Configuration::Ext::Agent::UnixSocket::ADAPTER
             "unix://#{agent_settings.uds_path}"
```

Comment:
Looking at this makes me think that perhaps all of this class is in the wrong location.

What you think of maybe adding this into the `AgentSettingsResolver` as a `agent_settings.url`? That way we'd have a centralized place to generate the correct url, rather than needing to repeat (and potentially get wrong) this code.

Reply:
I haven't looked into the details but at surface level I agree with what you are saying here that this logic seems like it should be elsewhere. The approval I had for repairing the crash was for a limited-time engagement, and refactoring working code - while good - wouldn't fall under that approval. Let's make a follow-up ticket to deal with this in the future?

Reply:
@ivoanjo Since you are on PTO let's connect on this when you are back.